[master] Use token's eauth key if load omits it.

RyanFrantz opened the pull request:
### What does this PR do?

- For token-based authentication, where administrators would also like to write a custom eauth ACL module, this update allows `__get_acl` to check for such a module, either via the `eauth_acl_module` configuration value or fallback to the `eauth` module's `acl` function, if defined.

### What issues does this PR fix or reference?

N/A

### Previous Behavior

If `load["eauth"]` is not defined, `__get_acl()` immediately returns `None`.

### New Behavior

In `get_auth_list()`, if the `"eauth"` key is not present in the `load` dictionary, but _is_ present in the `token` dictionary, the token's `"eauth"` value is added to the `load` dictionary. When `__get_acl()` is called from within `get_auth_list()`, the former can attempt to perform an ACL lookup via the `eauth_acl_module` configuration option or fall back to the eauth module's `acl()` function, if defined.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices, including the
[PR Guidelines](https://docs.saltproject.io/en/master/topics/development/pull_requests.html).

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

Commit:
Use token's eauth key if load omits it.

- `get_auth_list` uses the load to perform ACL lookup but if no `eauth`
  key is present, `__get_acl` immediately returns.
- For token-based authentication, where administrators would also like
  to write a custom eauth ACL module, this update allows `__get_acl` to
  check for such a module, either via the `eauth_acl_module`
  configuration value or fallback to the `eauth` module's `acl`
  function, if defined.

Akm0d reviewed (changes requested):
Needs a test

RyanFrantz committed:
Add tests

- Includes two tests:
  - Validates that the fallback to checking the static `external_auth` option works as expected.
  - Validates the case where a server is configured to perform external ACL lookups via `eauth_acl_module`, `load["eauth"]` is undefined and `token["eauth"]` _is_ defined.
- Verified via `nox -e 'test-3(coverage=False)' -- tests/pytests/unit/auth/test_auth.py`

RyanFrantz commented:
> Needs a test

I've added tests.

dwoz reviewed (changes requested):
There is a missing import in your test:

` from tests.support.mock import patch`

Also, after the import is added the test passes with and without your change. The test should make sure it's testing the functionality you are adding. 

You are using eauth to look up the `auth_list` before `eauth` has been validated against `external_auth` config. That's no good IMO because someone could leverage an eauth that backend should not be allowed based on the `external_auth` option.

With that being said. I think there should be an additional test to validate `auth_list` will only come from the token if the eauth backend has been configured.

This is a very sensitive part of the code base so we need to be critical of this change.

RyanFrantz committed:
Restores missing import

Not sure how I omitted that before. I think I cleaned it up by accident, along with some temporary comments I had left myself.

RyanFrantz commented:
> There is a missing import in your test:
> 
> ` from tests.support.mock import patch`

I'm fixing that up.

> 
> Also, after the import is added the test passes with and without your change. The test should make sure it's testing the functionality you are adding.

I chose to add two tests:

1. One demonstrating the expected behavior in this module, where a user's permissions are looked up via the static `external_auth` configuration option. I did this because there was no preexisting test and I thought it would be helpful to illustrate that case.
2. One demonstrating the functionality I added that enables external authorization to be performed, in the case where the `load` dictionary does not contain an `eauth` key, but it is present in the `token` dictionary.

> 
> You are using eauth to look up the `auth_list` before `eauth` has been validated against `external_auth` config. That's no good IMO because someone could leverage an eauth that backend should not be allowed based on the `external_auth` option.

I do not intend to subvert existing controls. I am trying to use the support in `__get_acl()` to perform external auth_list lookups via the `eauth_acl_module` option. If that is not defined or fails to generate a list, `__get_acl()` calls its default behavior of looking at the `external_auth` option dictionary to find an auth_list.

> 
> With that being said. I think there should be an additional test to validate `auth_list` will only come from the token if the eauth backend has been configured.

I'll look into this.

> 
> This is a very sensitive part of the code base so we need to be critical of this change.

I agree. And I welcome the scrutiny.

dmurphy18 reviewed (changes requested):
Needs changelog

Commit:
Add changelog

Commit:
Reformat test with black.

RyanFrantz commented:
@dmurphy18 @dwoz Do either of you need anything else for this PR to be approved?

dwoz commented:
@RyanFrantz Please rebase this.

RyanFrantz commented:
Closing this PR. It's been two years and the original project I was working on no longer exists.
